Add ConfirmView utility

TrustyJAID opened the pull request:
closes #6174

### Description of the changes



### Have the changes in this PR been tested?

<!--
Choose one (remove the line that doesn't apply):
-->
Yes
<!--
If the question doesn't apply (for example, it's not a code change), choose Yes.

Please respond to this question truthfully. We do not delay nor reject PRs
based on the answer to this question but it allows to better review this PR.
-->

TrustyJAID committed:
Add ConfirmView utility
closes #6174

TrustyJAID committed:
export the new view

flaree reviewed (commented):
Would it be worth adding an on_timeout that sets the buttons to disabled?

TrustyJAID commented:
I am open to customizing the timeout but I will not be adding support for disabling the buttons automatically. That would require sending the message through the view which I feel takes away from the simplicity of this. The buttons when interacted with will remove the buttons from the message but timeout will do nothing and it should be up to the user to handle that I think.

flaree commented:
Huh, I have a confirm that doesn't send through a view and works fine.

```py
async def on_timeout(self):
        for item in self.children:
            item.disabled = True
```

TrustyJAID commented:
That would disable the button and prevent any response but would still require editing the original message which this does not have any reference to.

TrustyJAID committed:
Add timeout kwarg and explicitly prevent this from being a persistent view

TrustyJAID committed:
Fix docs

TrustyJAID committed:
remove default arg and set result to `None` so we can know if the view was interacted with or not allowing for easier cleanup.
Add examples of cleanup including how to disable the buttons if they're not pressed.

TrustyJAID commented:
I have reworked this slightly so we can know if the buttons were not pressed so users can clean them up however they wish, whether that be disabling the buttons or removing them after the timeout.

TrustyJAID committed:
Fix docs

Jackenmen commented:
It shouldn't be required (i.e. we should only try doing it if not `None`) but I think that it would be simple enough to allow setting `view.message` attribute after the message is sent and remove the view from the message `on_timeout` in such a case. I would also suggest updating the relevant example to use that, i.e.
```py
view = ConfirmView(ctx.author)
view.message = await ctx.send("Are you sure you about that?", view=view)
await view.wait()
if view.result:
    await ctx.send("Okay I will do that.")
else:
    await ctx.send("I will not be doing that then.")
```

TrustyJAID committed:
Add `disable_buttons` kwarg that optionally disables the buttons instead of removing them
Move all logic for removal/disabling into `on_timeout` and update examples about setting a `view.message` attribute when sending the view as per Jack's suggestion.

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+
+    async def on_timeout(self):
+        if self.message is None:
+            return
+            # we can't do anything here if message is none
```
That's a weird order, comments generally are placed before the action they talk about:
```suggestion
            # we can't do anything here if message is none
            return
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+            return
+            # we can't do anything here if message is none
+        if self.disable_buttons:
+            # prioritize disabling buttons since it's considered custom
+            # behavior
```
I don't know what this is supposed to convey, making it `if not self.disable_buttons` with an early return would have been no different so I don't see any prioritization here.

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+        self.yes_button.disabled = True
+        self.no_button.disabled = True
+
+    @discord.ui.button(label=_("Yes"), style=discord.ButtonStyle.green)
+    async def yes_button(self, interaction: discord.Interaction, button: discord.ui.Button):
+        self.result = True
+        self.stop()
+
+    @discord.ui.button(label=_("No"), style=discord.ButtonStyle.red)
+    async def no_button(self, interaction: discord.Interaction, button: discord.ui.Button):
+        self.result = False
+        self.stop()
```
These have been changed to no longer remove the view from the message, why is that?

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+    Using the view::
+
+        view = ConfirmView(ctx.author)
+        view.message = await ctx.send("Are you sure you about that?", view=view)
+        # attach the message to the view this way to ensure clean removal
+        # when the view is finished.
```
That's a weird order, comments generally are placed before the action they talk about. 
The comment also seemed a bit unclear so here's a suggested rephrase:
```suggestion
        # Attach the message to the view after sending it.
        # This way, the view will be automatically removed
        # from the message after the timeout.
        view.message = await ctx.send("Are you sure you about that?", view=view)
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+    timeout: float
+        The timeout of the view in seconds. Defaults to ``180`` seconds.
+    disable_buttons: bool
+        Whether to disable the buttons after the timeout. Defaults to ``False``.
```
It would probably make sense to explain the default behavior:
```suggestion
        Whether to disable the buttons instead of removing them from the message after the timeout.
        Defaults to ``False``.
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+        sending the message. This can also be left as ``None`` in which case
+        if the view is never interacted with nothing will happen in ``on_timeout``
+    disable_buttons: bool
+        Whether to disable the buttons instead of remove the buttons if the message
+        has been seen by the view.
```
```suggestion
        Whether to disable the buttons instead of removing them on timeout
        (if the ``message`` attribute has been set on the view).
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+    message: Optional[discord.Message]
+        The message the confirm view is sent on. This can be set while
+        sending the message. This can also be left as ``None`` in which case
+        if the view is never interacted with nothing will happen in ``on_timeout``
```
This may be unnecessary but the sentence read weirdly to me:
```suggestion
        nothing will happen in ``on_timeout``, if the view is never interacted with.
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+            return
+        await self.message.edit(view=None)
+
+    def disable(self):
+        self.yes_button.disabled = True
+        self.no_button.disabled = True
```
Is this meant to be a public method? If so, please document it, otherwise prefix it with `_`.

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,105 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+        else:
+            await ctx.send("I will not be doing that then.")
+
+    Disabling if the buttons are not pressed::
```
```suggestion
    Auto-disable the buttons after timeout if nothing is pressed::
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -11,7 +11,7 @@
     from redbot.core.commands import Context
 
-__all__ = ("SimpleMenu", "SetApiModal", "SetApiView")
+__all__ = ("ConfirmView", "SimpleMenu", "SetApiModal", "SetApiView")
```
nitpick: we try to keep items in `__all__` in definition order to make it somewhat easier to check if anything is missing in `__all__`:
```suggestion
__all__ = ("SimpleMenu", "SetApiModal", "SetApiView", "ConfirmView")
```

TrustyJAID committed:
Clarify comments
remove disable() method
Actually respond to the interaction and call the `on_timeout` method if a button is pressed and we have stopped the view

TrustyJAID committed:
Reduce the colour noise of the buttons by making the no button grey.
Rename the button methods to more accurately describe their intended use `confirm_button` and `dismiss_button`.
Add docstrings explaining how one might customize the style of the `confirm/dismiss` buttons

TrustyJAID committed:
Show example of also modifying the label

TrustyJAID committed:
Fix typo in button examples

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+        await self.on_timeout()
+
+    async def interaction_check(self, interaction: discord.Interaction):
+        if self.message is None:
+            self.message = interaction.message
```
Looking at the documentation, it seems that it's consistent with it but I do wonder if we actually want to make the behavior of the view's timeout action dependent on whether some random person other than the author decides to click one of the buttons under the message? Would it not make more sense to this API's consumers to expect that not setting `view.message` will fully prevent it from removing/disabling the buttons on the message?

TrustyJAID replied:
I suppose my original thinking was to prefer cleanup of the view in as many situations as possible. For example, a cog sends the view without explicitly setting the message attribute, if the issuer of the command doesn't click the button but someone else does it would be better to then cleanup the view even if it doesn't actually respond to them. I view it as better in that the buttons no longer sit around unusable for anyone. I could move it under the author check to be more consistent and keep everything else the same I suppose. Though now I also wonder if I should be responding to someone who isn't the author with a message as well.

Jackenmen replied:
> Though now I also wonder if I should be responding to someone who isn't the author with a message as well.

I haven't thought about it but it would definitely make it seem not broken to bystanders. And it would be more consistent with all the other views/modals that we already have (just the simple menu and set api so far).

> I could move it under the author check to be more consistent and keep everything else the same I suppose.

It seems like an alright way of handling it. But with your explanation, I honestly am not sure if we should do that or just leave it as is (since I guess that the option of not trying to do the cleanup at all is not something we're interested in).

TrustyJAID replied:
I added a response if someone other than the author attempts to interact with the buttons. I left in setting the message object from any interaction though. I am much happier with the way this would work now and I would rather prioritize removing or disabling the view in as many situations as possible. Leaving broken buttons lying around outside of development and testing just feels bad.

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+        The timeout of the view in seconds. Defaults to ``180`` seconds.
+    disable_buttons: bool
+        Whether to disable the buttons instead of removing them from the message after the timeout.
+         Defaults to ``False``.
```
This may seem like a nitpick but this is actually a very important fix to how Sphinx ends up displaying this:
```suggestion
        Defaults to ``False``.
```
![image](https://github.com/Cog-Creators/Red-DiscordBot/assets/6032823/9d54f98b-1810-4f20-87a8-19fdaed9fc42)

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+
+class ConfirmView(discord.ui.View):
+    """
+    A simple ``discord.ui.View`` used for confirming something.
```
```suggestion
    A simple `discord.ui.View` used for confirming something.
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+    message: Optional[discord.Message]
+        The message the confirm view is sent on. This can be set while
+        sending the message. This can also be left as ``None`` in which case
+        nothing will happen in ``on_timeout``, if the view is never interacted with.
```
```suggestion
        nothing will happen in `on_timeout()`, if the view is never interacted with.
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+    @discord.ui.button(label=_("Yes"), style=discord.ButtonStyle.green)
+    async def confirm_button(self, interaction: discord.Interaction, button: discord.ui.Button):
+        """
+        A ``discord.ui.Button`` to confirm the message.
```
```suggestion
        A `discord.ui.Button` to confirm the message.
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+
+        Examples
+        --------
+        Changing the style and label of this ``discord.ui.Button``::
```
```suggestion
        Changing the style and label of this `discord.ui.Button`::
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+    @discord.ui.button(label=_("No"), style=discord.ButtonStyle.secondary)
+    async def dismiss_button(self, interaction: discord.Interaction, button: discord.ui.Button):
+        """
+        A ``discord.ui.Button`` to dismiss the message.
```
```suggestion
        A `discord.ui.Button` to dismiss the message.
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+
+        Examples
+        --------
+        Changing the style and label of this ``discord.ui.Button``::
```
```suggestion
        Changing the style and label of this `discord.ui.Button`::
```

Jackenmen commented on redbot/core/utils/views.py:
```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+        nothing will happen in ``on_timeout``, if the view is never interacted with.
+    disable_buttons: bool
+        Whether to disable the buttons isntead of removing them on timeout
+        (if the ``message`` attribute has been set on the view).
```
```suggestion
        (if the `message` attribute has been set on the view).
```

Jackenmen reviewed (changes requested):
There's mostly a bunch of docstring fixes left but I also commented with one implementation question.

TrustyJAID committed:
Apply suggestions from code review

Co-authored-by: Jakub Kuczys <me@jacken.men>

TrustyJAID committed:
Respond to users who are not set as the author with a response so the buttons don't appear broken

Jackenmen committed:
Add docstrings to on_timeout() and interaction_check()

Jackenmen committed:
Break line in example to obey line length limit

Jackenmen committed:
Examples->Example since there is just one

Jackenmen committed:
Explain what the button's callback does

Jackenmen committed:
Document confirm_button and dismiss_button as attributes, not methods

Jackenmen reviewed (approved):
This new utility works as expected so from my side, this is good to go.

I would, however, like you to look at the rendered documentation and assess whether the current version of it is acceptable:
https://red-discordbot--6176.org.readthedocs.build/en/6176/framework_utils.html#redbot.core.utils.views.ConfirmView

Basically, I looked at the rendered documentation and thought that showing the documentation from the base class on our overridden version of the methods was confusing and have instead opted to document it with the actual behavior of them. Additionally, I documented `confirm/dismiss_button` as a regular attribute of the `Button` type since that's what it actually ends up being on the instance of this view.

If everything seems fine to you, feel free to just merge this.
